solr/core/src/java/org/apache/solr/search/StringPayloadValueSource.java

```diff
@@ -0,0 +1,306 @@
+
+        if (terms != null) {
+          final TermsEnum termsEnum = terms.iterator();
+          if (termsEnum.seekExact(indexedBytes)) {
+            docs = termsEnum.postings(null, PostingsEnum.ALL);
+          } else {
+            docs = null;
+          }
+        } else {
+          docs = null;
+        }
+
+        if (docs == null) {
+          // dummy PostingsEnum so floatVal() can work
+          // when would this be called?  if field/val did not match?  this is called for every doc?  create once and cache?
+          docs = new PostingsEnum() {
+            @Override
+            public int freq() {
+              return 0;
+            }
+
+            @Override
+            public int nextPosition() throws IOException {
+              return -1;
+            }
+
+            @Override
+            public int startOffset() throws IOException {
+              return -1;
+            }
+
+            @Override
+            public int endOffset() throws IOException {
+              return -1;
+            }
+
+            @Override
+            public BytesRef getPayload() throws IOException {
+              return null;
+            }
+
+            @Override
+            public int docID() {
+              return DocIdSetIterator.NO_MORE_DOCS;
+            }
+
+            @Override
+            public int nextDoc() {
+              return DocIdSetIterator.NO_MORE_DOCS;
+            }
+
+            @Override
+            public int advance(int target) {
+              return DocIdSetIterator.NO_MORE_DOCS;
+            }
+
+            @Override
+            public long cost() {
+              return 0;
+            }
+          };
+        }
+        atDoc = -1;
+      }
+
+      @Override
+      public String strVal(int doc) {
+        try {
+          if (doc < lastDocRequested) {
+            // out-of-order access.... reset
+            reset();
+          } else if (doc == lastDocRequested) {
+            return docValue;
+          }
+
+          lastDocRequested = doc;
+
+          if (atDoc < doc) {
+            atDoc = docs.advance(doc);
+          }
+
+          if (atDoc > doc) {
+            // term doesn't match this document... either because we hit the
+            // end, or because the next doc is after this doc.
+            docValue = defaultValues.strVal(doc);
+            return docValue;
+          }
+
+          // a match!
+          int freq = docs.freq();
+          for (int i = 0; i < freq; i++) {
+            docs.nextPosition();
+            BytesRef payload = docs.getPayload();
+            if (payload != null) {
+              return bytesRefToString(payload);
+            }
+          }
+          docValue = defaultValues.strVal(doc);
+          return docValue;
+        } catch (IOException e) {
+          throw new RuntimeException("caught exception in function " + description() + " : doc=" + doc, e);
+        }
+      }
+    };
+  }
+
+  private String bytesRefToString(BytesRef payload) {
```

Comment:
I suppose it should be just bytesRef.utf8ToString()

Reply:
right, removed.